### PR TITLE
ColorPickerPopover: apply react-aria for a11y support

### PR DESCRIPTION
**Pull request by thisisobate**
**What this PR does / why we need it**:
This PR adds focus trapping to ColorPickerPopover component.
**Which issue(s) this PR fixes**:

Relates to #https://github.com/grafana/grafana/issues/42372

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by thisisobate** [comment]
Seems like pressing the `esc` key in panelEdit closes the panelEdit window. This would hinder us from using `esc` in ColorPickerPopover because it is rendered within PanelEdit.

**Comment by thisisobate** [comment]
Currently exploring different ways I could solve the `esc` issue.

File: packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx [hunk 1]
```diff
@@ -117,16 +118,18 @@ class UnThemedColorPickerPopover<T extends CustomPickersDescriptor> extends Reac
...
       <div className={styles.colorPickerPopover}>
-        <div className={styles.colorPickerPopoverTabs}>
-          <div className={this.getTabClassName('palette')} onClick={this.onTabChange('palette')}>
-            Colors
-          </div>
-          <div className={this.getTabClassName('spectrum')} onClick={this.onTabChange('spectrum')}>
-            Custom
+        <FocusScope contain restoreFocus autoFocus>
+          <div className={styles.colorPickerPopoverTabs}>
+            <div tabIndex={0} className={this.getTabClassName('palette')} onClick={this.onTabChange('palette')}>
```
Comment (thisisobate) [inline packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:123]: Considering using button here instead.

**Comment by thisisobate** [comment]
Figured the root cause of the `esc` key problem. 

**Some context**: Adding FocusScope to colorpickerpopover emits an event which bubbles up to the parent (PanelEdit). Luckily, FocusScope prevents event from bubbling.

Now, the problem was that I was using FocusScope in a class component without knowing that the `esc` functionality is being handled by a `useRestoreFocus` hook under the hood.

So, converting the colorpickerpopover component from class to a function would solve the `esc` issue.

The only problem we would encounter now would be that focus won't be completely trapped within the colorpickerpopover component. Reason is because colorPickerPopover is wrapped within the popper component. We can solve this by either by appending another focusScope above colorPickerPopover component, or append `<span data-focus-scope-start="true" hidden="true">` in the beginning, and `<span data-focus-scope-end="true" hidden="true">` at the end within the popover component.

File: packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx (hunk 1 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:123]: Have you looked at this @thisisobate https://github.com/adobe/react-spectrum/blob/1e2b7f280a04ead83e21b27fa580eb189e5f4186/packages/%40react-aria/overlays/docs/DismissButton.mdx
We're using it here in the NavBar: https://github.com/grafana/grafana/blob/0611e435fc976e319ae71c9becbb5130feb6ec1d/public/app/core/components/NavBar/NavBarItemMenuTrigger.tsx#L140

**Comment by hugohaggmark** [comment]
> So, converting the colorpickerpopover component from class to a function would solve the `esc` issue.

How would converting the ColorPickerPopover component from class to function solve the `esc` issue?

> The only problem we would encounter now would be that focus won't be completely trapped within the colorpickerpopover component. Reason is because colorPickerPopover is wrapped within the popper component. We can solve this by either by appending another focusScope above colorPickerPopover component, or append `<span data-focus-scope-start="true" hidden="true">` in the beginning, and `<span data-focus-scope-end="true" hidden="true">` at the end within the popover component.

Do I understand this correctly that converting the ColorPickerPopover to functional component solves one challenge but introduces this new challenge with the focus trap?

**Comment by thisisobate** [comment]
> How would converting the ColorPickerPopover component from class to function solve the esc issue?

FocusScope, which is a react-aria utility, works completely in a functional component because it uses some custom hooks under the hood. In the case of the `esc` issue, useRestoreFocus hooks which is used in FocusScope takes care of that.

**Comment by thisisobate** [comment]
> Do I understand this correctly that converting the ColorPickerPopover to functional component solves one challenge but introduces this new challenge with the focus trap?

No. The focus trap issue would always exist regardless of whether or not I convert the ColorPickerPopover to a functional component.

FocusScope identifies it's scope by it's starting and ending tags. Since ColorPickerPopover is used within the Popover component, focus trapping won't completely trap within the ColorPickerPopover. For this to work, we have to wrap FocusScope around both ColorPickerPopover and Popover components.

**Comment by thisisobate** [comment]
Surprisingly, the issue of focus not completely trapping within the ColorPickerPopover component was as a result of listening to the `KeyDown` event. Would use the keyUp event instead.

**Comment by thisisobate** [comment]
**New Findings**
Figured the focus escape occurring in ColorPickerPopover was as a result of not handling the tab key in the onKeyDown event handler.

File: packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx (hunk 1 above)
Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:123]: Haven't explored this approach yet. Would take it for a spin

Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:123]: DismissButton doesn't seem to solve the `esc` issue in panelEdit.
I think my current solution already solves it.

Comment (hugohaggmark) [reply packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:123]: The DismissButton is for people that use screen readers and might not default to using the `esc` key.

File: packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx [hunk 2]
```diff
@@ -38,6 +38,20 @@ export const colorPickerFactory = <T extends ColorPickerProps>(
...
+
+      event.stopPropagation();
+
+      if (event.key === 'Escape') {
+        hidePopper();
+      }
+
+      return;
+    };
+
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx:54]: I think I would wait with this and have this PR only solve the focus trap. In later PRs where you need to adress the keyboard navigation you can revisit this.

**Review by hugohaggmark — APPROVED** [review]
Looks good but I think you should tackle the keyboard navigation entirely in a separate PR and not small pieces of it in this.

**Pull request merged**